src/rosdep2/installers.py

```diff
@@ -368,7 +368,10 @@ def get_packages_to_install(self, resolved, reinstall=False):
         if not resolved:
             return []
         else:
-            return [x for x in resolved if x not in self.detect_fn(resolved)]
+            detected = set(self.detect_fn(resolved))
+            seen = set()
+            seen_add = seen.add
+            return [x for x in resolved if (x not in detected) and (x not in seen or seen_add(x))]
```

Comment:
What is the `seen` set for? Would this not be simpler:

```python
detected = self.detect_fn(resolved)
return [x for x in resolved if x not in detected]
```

Reply:
It's fine, I just wasn't aware of why. You might want a comment, otherwise someone might just remove it as redundant.

Reply:
Yes, it would appear I overthought the problem.